File: backend/apps/shared/utils/trace.py
"""
Request tracing utilities
"""
import uuid
from typing import Optional
from contextvars import ContextVar

# Context variable for trace ID
_trace_id: ContextVar[Optional[str]] = ContextVar('trace_id', default=None)


def generate_trace_id() -> str:
    """
    Generate a unique trace ID
    
    Returns:
        Unique trace ID string
    """
    return str(uuid.uuid4())


def set_trace_id(trace_id: str):
    """
    Set trace ID in context
    
    Args:
        trace_id: Trace ID to set
    """
    _trace_id.set(trace_id)


def get_trace_id() -> Optional[str]:
    """
    Get trace ID from context
    
    Returns:
        Trace ID if set, None otherwise
    """
    return _trace_id.get()

File: ksit_nexus_app/lib/utils/trace.dart
/// Request tracing utilities for Flutter
import 'package:flutter/foundation.dart';

class TraceId {
  static final TraceId _instance = TraceId._internal();
  factory TraceId() => _instance;
  TraceId._internal();

  String? _currentTraceId;

  /// Generate a new trace ID
  String generate() {
    _currentTraceId = _generateTraceId();
    return _currentTraceId!;
  }

  /// Get the current trace ID
  String? get current => _currentTraceId;

  /// Set the trace ID (e.g., from response headers)
  void set(String traceId) {
    _currentTraceId = traceId;
  }

  /// Clear the current trace ID
  void clear() {
    _currentTraceId = null;
  }

  String _generateTraceId() {
    // Generate a UUID-like trace ID
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    final random = (timestamp * 1000 + (timestamp % 1000)).toString();
    return 'trace-$random-${DateTime.now().toIso8601String()}';
  }
}

/// Global trace ID instance
final traceId = TraceId();

